pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -103,13 +104,107 @@
  */
-export class Terminal extends React.Component {
-    constructor(props) {
+
+export class TerminalState {
+    terminal: Term;
+    element: HTMLDivElement;
+    channel: cockpit.Channel<string> | null;
+
+    #default_channel_creator: (() => cockpit.Channel<string>) | undefined;
+
+    constructor(default_channel_creator?: () => cockpit.Channel<string>) {
+        this.terminal = new Term({
+            cols: 1,
+            rows: 1,
+            cursorBlink: true,
+            fontSize: 16,
+            fontFamily: 'Menlo, Monaco, Consolas, monospace',
+            screenReaderMode: true,
+        });
+        this.terminal.loadAddon(new CanvasAddon());
+        this.element = document.createElement("div");
+        this.channel = null;
+        this.#default_channel_creator = default_channel_creator;
+    }
+
+    connectChannel(channel: cockpit.Channel<string> | undefined) {
+        if (!channel && !this.channel && this.#default_channel_creator)
+            channel = this.#default_channel_creator();
+        if (channel && channel != this.channel) {
+            this.disconnectChannel();
+            this.channel = channel;
+
+            channel.addEventListener('message', (_event, data) => {
+                this.terminal.write(data);
+            });
+
+            this.terminal.onData(function(data) {
+                if (channel.valid) {
+                    /* HACK: Ctrl+Space (and possibly other
+                     * characters) is a disaster: While it is U+00A0
+                     * in unicode, with an UTF-8 representation of
+                     * 0xC2A0, the "visible" string in JS is 0x00
+                     * (with TextEncoder, btoa(), and string
+                     * comparison). The internal representation
+                     * retains half of it, and trying to send it to
+                     * the websocket would result in a single 0xA0,
+                     * which is invalid UTF-8 (and causes the session
+                     * to crash). So intercept and ignore such broken
+                     * chars.  See
+                     * https://github.com/cockpit-project/cockpit/issues/21213 */
+                    if (data === '\x00') {
+                        console.log("terminal: ignoring invalid input", data);
+                        return;
+                    }
+                    channel.send(data);
+                }
+            });
+
+            channel.addEventListener('close', (_event, options) => {
+                const term = this.terminal;
+                term.write('\x1b[31m' + (options.problem || 'disconnected') + '\x1b[m\r\n');
+                term.refresh(term.rows, term.rows);
+            });
+        }
+    }
+
+    disconnectChannel() {
+        if (this.channel) {
+            this.channel.close();
+            this.channel = null;
+        }
+    }
+
+    close() {
+        this.disconnectChannel();
+        this.terminal.dispose();
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-5fcf8444-20250814-082605-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)